### PR TITLE
TT-85 fix: fix overlapping seconds on time entries

### PR DESCRIPTION
**Pull request by PaulRC-ioet**
fix a bug that not allow to update changes on "time entries".

### CODE REVIEW

File: src/app/modules/shared/components/details-fields/details-fields.component.spec.ts [hunk 1]
```diff
@@ -76,6 +76,11 @@ describe('DetailsFieldsComponent', () => {
     description: '',
     technology: '',
   };
+  const dateTest = moment().format('YYYY-MM-DD');
+  const endHourTest = moment().format('HH:mm:ss');
+  const endDateTest = new Date(`${dateTest}T${endHourTest.trim()}`);
+  const startHourTest = moment().subtract(3, 'hours').format('HH:mm:ss');
+  const startDateTest = new Date(`${dateTest}T${startHourTest.trim()}`);
```
Comment (josepato87) [inline src/app/modules/shared/components/details-fields/details-fields.component.spec.ts:83]: I would move these two lines before the `end` variables

Comment (josepato87) [inline src/app/modules/shared/components/details-fields/details-fields.component.spec.ts:83]: Do you think we could have any issue about the timing where the tests run? For example, if we put this test running at 1:00 am, due you are subtracting 3 hours, it means that the current hour would be 23:00 of what day?

File: src/app/modules/shared/components/details-fields/details-fields.component.spec.ts [hunk 2]
```diff
@@ -314,6 +319,19 @@ describe('DetailsFieldsComponent', () => {
...
+    component.entryToEdit = {...entryToEdit, description: 'test', };
+    fixture.componentInstance.ngOnChanges();
+
+    const startHourValue = moment().subtract(3, 'hours').format('HH:mm');
+    const endHourValue = moment().format('HH:mm');
+
+    component.onSubmit();
+
+    expect(component.starDateValue).toEqual(startHourValue);
+    expect(component.endDateValue).toEqual(endHourValue);
```
Comment (josepato87) [inline src/app/modules/shared/components/details-fields/details-fields.component.spec.ts:332]: How are you asserting that the other fields, like description are being updated?

File: src/app/modules/shared/components/details-fields/details-fields.component.spec.ts [hunk 3]
```diff
@@ -314,6 +319,19 @@ describe('DetailsFieldsComponent', () => {
     expect(component.saveEntry.emit).toHaveBeenCalledWith(data);
   });
 
+  it('onSubmit an entry without change hours, should not modify the start_date and end_date ', () => {
+    component.entryToEdit = {...entryToEdit, description: 'test', };
+    fixture.componentInstance.ngOnChanges();
+
+    const startHourValue = moment().subtract(3, 'hours').format('HH:mm');
+    const endHourValue = moment().format('HH:mm');
```
Comment (josepato87) [inline src/app/modules/shared/components/details-fields/details-fields.component.spec.ts:327]: not sure if this should be the best approach for setting the start and end hours values, as I understand, you are setting those values in the `before each` section. Could you read the current entry "hours" and setting the test variables with those values instead?

File: src/app/modules/shared/components/details-fields/details-fields.component.ts [hunk 4]
```diff
@@ -204,13 +208,20 @@ export class DetailsFieldsComponent implements OnChanges, OnInit {
     }
     const startDate = this.entryForm.value.start_date;
     const endDate = this.entryForm.value.end_date;
+    this.starDateValue = this.entryForm.value.start_hour === this.starDateValue ?
+      this.entryToEdit.start_date :
+      new Date(`${startDate}T${this.entryForm.value.start_hour.trim()}`).toISOString();
+    this.endDateValue = this.entryForm.value.end_hour === this.endDateValue ?
+      this.entryToEdit.end_date :
+      new Date(`${endDate}T${this.entryForm.value.end_hour.trim()}`).toISOString();
```
Comment (josepato87) [inline src/app/modules/shared/components/details-fields/details-fields.component.ts:216]: I see an opportunity to extract lines 213 and 216 into a method that receives the `entryForm.value.X_VALUE` and returns everything that is being done in that large sentence.

If you find that in other files, could be an utilitary. :)

File: src/app/modules/shared/components/details-fields/details-fields.component.spec.ts (hunk 1 above)
Comment (PaulRC-ioet) [reply src/app/modules/shared/components/details-fields/details-fields.component.spec.ts]: I have change all the tests, please check them.

Comment (PaulRC-ioet) [reply src/app/modules/shared/components/details-fields/details-fields.component.spec.ts]: I have deleted that block of lines, it's different now.

File: src/app/modules/shared/components/details-fields/details-fields.component.spec.ts (hunk 2 above)
Comment (PaulRC-ioet) [reply src/app/modules/shared/components/details-fields/details-fields.component.spec.ts]: there's a new test to prove that.

File: src/app/modules/shared/components/details-fields/details-fields.component.spec.ts (hunk 3 above)
Comment (PaulRC-ioet) [reply src/app/modules/shared/components/details-fields/details-fields.component.spec.ts]: I have also deleted those hours, now is in the test those values.

File: src/app/modules/shared/components/details-fields/details-fields.component.ts (hunk 4 above)
Comment (PaulRC-ioet) [reply src/app/modules/shared/components/details-fields/details-fields.component.ts]: It's Done thanks.

File: src/app/modules/shared/components/details-fields/details-fields.component.spec.ts [hunk 5]
```diff
@@ -310,13 +312,71 @@ describe('DetailsFieldsComponent', () => {
       },
       shouldRestartEntry: false
     };
-
     expect(component.saveEntry.emit).toHaveBeenCalledWith(data);
   });
 
+  it('should not modify the start_date when start_hour has not been modified', () => {
+    const dateTest = moment().format('YYYY-MM-DD');
```
Comment (scastillo-jp) [inline src/app/modules/shared/components/details-fields/details-fields.component.spec.ts:319]: change: 
 `dateTest` to `currentDate`
 `startHourTest` to `startHour`

I think is necessary use other name of variable per example  `dateTest` to `dateCurrent`

Comment (PaulRC-ioet) [reply src/app/modules/shared/components/details-fields/details-fields.component.spec.ts:319]: Ok it's changed

File: src/app/modules/shared/components/details-fields/details-fields.component.ts [hunk 6]
```diff
@@ -197,29 +197,56 @@ export class DetailsFieldsComponent implements OnChanges, OnInit {
...
     }
-    const isStartDateInTheFuture = moment(startDate).isAfter(moment());
-    const isEndDateInTheFuture = moment(endDate).isAfter(moment());
-    if (isStartDateInTheFuture || isEndDateInTheFuture) {
+
+    const isStartDateInTheFuture = moment(startDateToSubmit).isAfter(moment());
+    const isEndDateInTheFuture = moment(endDateToSubmit).isAfter(moment());
+    const timeEntryIsInTheFuture = isStartDateInTheFuture || isEndDateInTheFuture;
+
+    if (timeEntryIsInTheFuture) {
```
Comment (Angeluz-07) [inline src/app/modules/shared/components/details-fields/details-fields.component.ts:249]: I remember we made a change to extract `timeEntryIsInTheFuture` into a function. I guess that caused issues in existing tests and you ended up doing it like this. Am I right, or  is it an error?

Comment (PaulRC-ioet) [reply src/app/modules/shared/components/details-fields/details-fields.component.ts:249]: I just tried the function of not allowing a "time entry" in the future and it allowed me to add. It ignores the way we put it so I left it as it was. but with the tests, there was no problem.

Comment (Angeluz-07) [reply src/app/modules/shared/components/details-fields/details-fields.component.ts:249]: It's ok. Then, keep in mind that, when we may want to retake that little refactor.

**Review by josepato87 — APPROVED** [review]
LGTM

**Pull request merged**